Sources/CacheAdvance/FileHandleExtensions.swift

```diff
@@ -0,0 +1,160 @@
+        let startingOffset = offsetInFile
+        switch try nextEncodedMessageSpan(cacheOverwritesOldMessages: cacheOverwritesOldMessages) {
+        case let .endOfNewestMessageMarker(offsetOfNextMessage):
+            // We found the last message!
+            if shouldSeekToOldestMessageIfFound {
+                // Seek to it.
+                try seek(toOffset: offsetOfNextMessage)
+            } else {
+                // Seek back to where we started.
+                try seek(toOffset: startingOffset)
+            }
+            return false
+
+        case .emptyRead:
+            // The remaining file is empty.
+            if shouldSeekToOldestMessageIfFound {
+                // The oldest message is at the beginning of the file.
+                try seek(toOffset: 0)
+            } else {
+                // Seek back to where we started.
+                try seek(toOffset: startingOffset)
+            }
+            return false
+
+        case let .span(messageLength):
+            // There's a valid message here. Seek ahead of it.
+            try seek(toOffset: offsetInFile + UInt64(messageLength))
+            return true
+
+        case .invalidFormat:
+            throw CacheAdvanceReadError.fileCorrupted
+        }
+    }
+
+    // MARK: Private
+
+    /// Returns the next encoded message span, seeking to the end the span.
+    private func nextEncodedMessageSpan(cacheOverwritesOldMessages: Bool) throws -> NextMessageSpan {
+        let messageSizeData = try __readDataUp(toLength: Data.messageSpanStorageLength)
+
+        guard messageSizeData.count > 0 else {
+            // We haven't written anything to this file yet, or we've reached the end of the file.
+            return .emptyRead
+        }
+
+        guard messageSizeData != Data.endOfNewestMessageMarker else {
+            // We have reached the most recently written message.
+            if cacheOverwritesOldMessages {
+                // We have a span marking the offset of the oldest message.
+                let offsetOfFirstMessageData = try __readDataUp(toLength: Data.bytesStorageLength)
+                guard let offsetOfFirstMessage = Bytes(offsetOfFirstMessageData) else {
+                    // The file is improperly formatted.
+                    return .invalidFormat
+                }
+                return .endOfNewestMessageMarker(offsetOfFirstMessage: offsetOfFirstMessage)
+
+            } else {
+                // The first message is always at the beginning of the file.
+                return .endOfNewestMessageMarker(offsetOfFirstMessage: 0)
+            }
+        }
+
+        guard let messageSize = MessageSpan(messageSizeData) else {
+            // The file is improperly formatted.
+            return .invalidFormat
+        }
+
+        return .span(messageSize)
+    }
+
+}
+
+private enum NextMessageSpan {
+    case span(MessageSpan)
+    case emptyRead
+    case endOfNewestMessageMarker(offsetOfFirstMessage: Bytes)
```

Comment:
I wonder if we should say `offsetOfOldestMessage` to be more symmetric?

Reply:
great call. updated!